src/main/kotlin/org/pkl/lsp/completion/UnqualifiedAccessCompletionProvider.kt

```diff
@@ -239,48 +256,391 @@ class UnqualifiedAccessCompletionProvider(private val project: Project) : Comple
+            }
+          }
+        }
+
+        true // typed objects cannot have elements
+      }
+      else -> {
+        val thisClassType = thisType.toClassType(base, context) ?: return false
+        when {
+          thisClassType.classEquals(base.mappingType) -> {
+            addMappingCompletions(alreadyDefinedProperties, thisClassType, base, collector, context)
+            collector.addAll(DEFINITION_LEVEL_KEYWORD_ELEMENTS)
+            true // mappings cannot have elements
+          }
+          thisClassType.classEquals(base.listingType) -> {
+            addListingCompletions(alreadyDefinedProperties, thisClassType, base, collector, context)
+            collector.addAll(DEFINITION_LEVEL_KEYWORD_ELEMENTS)
+            false
+          }
+          thisClassType.classEquals(base.dynamicType) -> {
+            collector.addAll(DEFINITION_LEVEL_KEYWORD_ELEMENTS)
+            false
+          }
+          else -> false
+        }
+      }
+    }
+  }
+
+  private fun addTypedCompletions(
+    alreadyDefinedProperties: Set<String>,
+    isClassOrModule: Boolean,
+    thisType: Type,
+    base: PklBaseModule,
+    collector: MutableList<CompletionItem>,
+    context: PklProject?,
+  ) {
+
+    val properties =
+      when (thisType) {
+        is Type.Class -> thisType.ctx.cache(context).properties
+        is Type.Module -> thisType.ctx.cache(context).properties
+        else -> unexpectedType(thisType)
+      }
+
+    for ((propertyName, property) in properties) {
+      if (propertyName in alreadyDefinedProperties) continue
+      if (property.isFixedOrConst && !isClassOrModule) continue
+
+      val propertyType = property.type.toType(base, thisType.bindings, context)
+      val amendedPropertyType = propertyType.amended(base, context)
+      if (amendedPropertyType != Type.Nothing && amendedPropertyType != Type.Unknown) {
+        val amendingPropertyType = propertyType.amending(base, context)
+        collector += createPropertyAmendElement(propertyName, amendingPropertyType, property)
+      }
+      collector += createPropertyAssignElement(propertyName, propertyType, property)
+    }
+  }
+
+  private fun addMappingCompletions(
+    alreadyDefinedProperties: Set<String>,
+    thisType: Type.Class,
+    base: PklBaseModule,
+    collector: MutableList<CompletionItem>,
+    context: PklProject?,
+  ) {
+
+    val keyType = thisType.typeArguments[0]
+    val valueType = thisType.typeArguments[1]
+
+    val amendedValueType = valueType.amended(base, context)
+    if (amendedValueType != Type.Nothing && amendedValueType != Type.Unknown) {
+      val amendingValueType = valueType.amending(base, context)
+      if ("default" !in alreadyDefinedProperties) {
+        collector +=
+          createPropertyAmendElement("default", amendingValueType, base.mappingDefaultProperty)
+      }
+      collector += createEntryAmendElement(keyType, amendingValueType, base)
+    }
+    if ("default" !in alreadyDefinedProperties) {
```

Comment:
This smells off... do you want to overwrite the `amendingValueType`'s `default` here?

Reply:
Not sure what you mean by "overwrite the `amendingValueType`'s `default`".

This code will add a `default = ...` autocomplete suggestion, for the case where you want to overwrite the `Mapping`/`Listing` default.